### PR TITLE
Add save and load of HNSW index

### PR DESCRIPTION
**Pull request by arnej27959**
* split graph data out into a simpler struct
* implement HnswIndexSaver
* implement HnswIndexLoader
* add unit test checking that saving and loading
  works as expected

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe @toregge FYI - please take a look if you want

### CODE REVIEW

File: searchlib/src/vespa/searchlib/tensor/hnsw_index_loader.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    const uint32_t *ptr = static_cast<const uint32_t *>(buf.buffer());
+    const uint32_t *end = ptr + num_readable;
+    uint32_t entry_docid = *ptr++;
+    int32_t entry_level = *ptr++;
+    uint32_t num_nodes = *ptr++;
+    std::vector<uint32_t> link_array;
+    for (uint32_t docid = 0; docid < num_nodes; ++docid) {
+        if (ptr == end) return false;
+        uint32_t num_levels = *ptr++;
+        if (num_levels > 0) {
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index_loader.cpp]: If num_levels is 0 (representing a removed document), we should create an entry for this document to ensure that node_refs has the right size. If the last document loaded has 0 levels (removed) the size of node_refs will be off by one.

File: searchlib/src/vespa/searchlib/tensor/hnsw_index_saver.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "nearest_neighbor_index_saver.h"
+#include <vespa/vespalib/datastore/entryref.h>
+#include <vector>
+
+namespace search::tensor {
+
+class HnswGraph;
+
+class HnswIndexSaver : public NearestNeighborIndexSaver {
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index_saver.h]: Please add class description.

**Review by geirst — COMMENTED** [review]
Very good 👍 Some comments to consider.

**Comment by arnej27959** [comment]
@geirst PTAL

**Review by geirst — APPROVED** [review]
👍

File: searchlib/src/vespa/searchlib/tensor/hnsw_index_loader.h [hunk 3]
```diff
@@ -2,19 +2,34 @@
...
 public:
     HnswIndexLoader(HnswGraph &graph);
     ~HnswIndexLoader();
     bool load(const fileutil::LoadedBuffer& buf);
 private:
     HnswGraph &_graph;
+    const uint32_t *_ptr;
+    const uint32_t *_end;
+    bool _failed;
+    uint32_t nextVal() {
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index_loader.h]: Please consider snake case for consistency.

**Pull request merged**